ext/socket/raddrinfo.c

```diff
@@ -318,17 +320,42 @@ rb_getaddrinfo(const char *node, const char *service,
     if (ret == 0)
         allocated_by_malloc = 1;
     else {
+        VALUE scheduler = rb_scheduler_current();
```

Comment:
This is `Addrinfo.getaddrinfo` function call chain:
- `addrinfo_s_getaddrinfo`
- `addrinfo_list_new`
- `call_getaddrinfo`
- `rsock_getaddrinfo`
- `rb_getaddrinfo` 👈 we are adding a hook here
- `nogvl_getaddrinfo`
- `getaddrinfo` syscall

A hook is added so "deep" because:
- Other ruby methods are calling deep into the above function chain. For example `Socket.getnameinfo` is calling `rb_getaddrinfo` directly.
- `rb_getaddrinfo` function is in `ext/socket/rubysocket.h`. So, can `rb_getaddrinfo` be used in ruby extensions? If yes, we're covered: extension function calls will also be delegated to a Scheduler hook.

Reply:
``` c
// Option 1:
// add this interface to core ruby
rb_address_resolve(...) # not in an extension
// rb_getaddrinfo -> calls rb_address_resolve if it's blocking lookup.

// Option 2:
rb_getaddrinfo -> core ruby
-> invoke the hook

// Option 3:
// expose rb_scheduler functions
```